WeakAuras/Modernize.lua

```diff
@@ -1315,81 +1315,45 @@ function Private.Modernize(data)
                 if subRegionIndex and property then
-                  change.property = "sub." .. subRegionIndex + 1 .. "." .. property
+                  if (afterPos and subRegionIndex > afterPos) and (beforePos and subRegionIndex < beforePos) then
+                    change.property = "sub." .. subRegionIndex + (offset or 1) .. "." .. property
+                  end
                 end
               end
             end
           end
         end
       end
     end
-
-    if data.regionType == "aurabar" then
-      data.subRegions = data.subRegions or {}
-      -- rename aurabar_bar into subforeground, and subbarmodel into submodel
-      for index, subRegionData in ipairs(data.subRegions) do
-        if subRegionData.type == "aurabar_bar" then
-          subRegionData.type = "subforeground"
-        elseif subRegionData.type == "subbarmodel" then
-          subRegionData.type = "submodel"
-        end
-        if subRegionData.bar_model_visible ~= nil then
-          subRegionData.model_visible = subRegionData.bar_model_visible
-          subRegionData.bar_model_visible = nil
-        end
-        if subRegionData.bar_model_alpha ~= nil then
-          subRegionData.model_alpha = subRegionData.bar_model_alpha
-          subRegionData.bar_model_alpha = nil
-        end
-      end
-      -- check if "subforeground" exists, add it if not
-      local found = false
+    local function check_and_fix_subregion(data, subregionType)
+      -- search existing ones
+      local indexes = {}
       for index, subRegionData in ipairs(data.subRegions) do
-        if subRegionData.type == "subforeground" then
-          found = true
+        if subRegionData.type == subregionType then
+          table.insert(indexes, index)
         end
       end
-      if not found then
+      -- add if missing
+      if #indexes == 0 then
         tinsert(data.subRegions, 1, {
-          ["type"] = "subforeground"
+          ["type"] = subregionType
         })
-        bump_condition_subregions(data)
-      end
-    end
-    -- rename conditions for bar_model_visible and bar_model_alpha
-    if data.conditions then
-      for conditionIndex, condition in ipairs(data.conditions) do
-        if type(condition.changes) == "table" then
-          for changeIndex, change in ipairs(condition.changes) do
-            if change.property then
-              local prefix, property = change.property:match("(sub%.%d+%.)(.*)")
-              if prefix and property then
-                if property == "bar_model_visible" then
-                  change.property = prefix.."model_visible"
-                elseif property == "bar_model_alpha" then
-                  change.property = prefix.."model_alpha"
-                end
-              end
-            end
-          end
+        move_condition_subregions(data, 1)
+      -- delete duplicate
+      elseif #indexes > 1 then
+        for i = 2, #indexes do
+          table.remove(data.subRegions, i)
```

Comment:
That line is incorrect, you don't want to remove all sub regions from 2 to #indexes. You want to remove indexes[i]. That is you do want to use the values stored in indexes

I have prepared a few auras for you.
First the uninteresting starting point, auras without subelements:
!WA:2!DAvxtTTrx44XtNu0KMc0uAZhVjU0jmKMMuW0qAEBtBSm2btm2UYciPHzCwPDT1gKLeRwbbURE6KHR5NGVox5Fc8l4mm9xa)e4UExpRKiazAsFF5cK3D1EwDopppNNLmlpAy(5yBmA3rPJs35A7BrSxJk8dk676lwiB2SnYo1DV3bw(cktON(YGH1D5BVnrqZz677k5bgH2exwwRK95i2QE72HmzMbepBhFrdFUN0QyPAMLm6NSNM8TzF0GKXjFSmzYKTV03EdMiK779LzA9xIxLEocITexlCwJqjri1SAZ94HoA64psTEsbVthmSRmHiD4R7tzwrTBBUvatmFPQnkVuvDjoXGeji57fgWCDRqd1oimYITbZt2e3n)vdAvSqtZwnnlyyQhHhUvGlzlMWWJ0LfQzeVvR5zexPZBdTHGHHA0SrPQv3nYlnf0sR0s0oSGrAUEerWYvoY1n3koCjBVK3MuGd1xW6GfOkDn6i8Jc2nK52og3GHgNt3TgBZCpw9clBFV28oAdAlWuQPuqKKShMGZLvlPodRMfnkvQ2aUxBFrxIc70sb7kE43l7(Kijgq9Gyyv7q87WeEe3Lta)NVlXJNe3SWNwEpgju9PyEDKoNhgt3Z3JTpfrs1oAPWvridtmAyp1wvPa896DjCVYWmyaW3d3dMfUp(7WV7k9vXXtYdDxFcTSLeftEYXm6gHIlnDKSShd(jndBxsyOAKEikFudYhXPdCNmA9BDRkr2pXyu7f2z8(C8iB6hjSzx7CWNKboFg4JHH23gfRu)n9AUjpG9I9oAQIGC0vX8cy0lob8zVgUe85yk(fWxQbxgUcCv4AW)rdUoCdih8vW4AbJOeKBWmtO6f9PS3Coy6NF4Amwqbv(knuyJdmjmekfdjsfyXC6JAgJyQouZYHX74i)D4)wg(rvXapm(5ph)8xIFobuYytov687qEft8UO5p8VJVWD10323VBg4I4jObdRbJObxq1FHcR8OWYW(OU9SWTZcFlCh4rdM2R0gDZp1lPtxd(gnb3t1xYCcU0rWMj7vY54Helxg1XG4g4qYcpqR3rV3bMsBNVRFyarS2kQs4ZbDKos5I4Ln8LXkOmbJCQ5k8uVWsM1Hso)Vsjq53CUyIV1jQMPFUIa6zrs9yYIEmW8o7f)1sA9EwgOYvgdgBGYya3yl8pyHB8biLdIJwhLOuvIMVWCZfmCuiRa9LrHsgLrxK8kNbX7AEoLY8mQvA5sgNDg8ee37(L4iopHin3rT(jlTNMz3yajKtzggvE88MitMSbf5fjybtwr143MyZwTijuY96OteRUuL7C8S70ufqA5eRyFOqMeEVJUha1lR4AgnB31Bp3sVeQMXiE7oVvMzO8VYdp5pUayKOWUD2ePrXJzRgkjZaFbhB(J1b9NVUrLFREnZcvrLfkKpuD)KYF0J(2lnIVHANBIlkD5EOJ6sMvRuRuWTP8qL5TQuBLyd2kOLK3L1IUfANZTBjDeSqhuUU1(Hoeu0(SeGdLPOgni37)esMLjyyBKi87Q2XsbuSdhU(bO1QVY70ow1RxaNEQwV)pe0bx9FmfsgyOYekQX7Tj6SVIGe0BL0b6T99KbdxwW3o3VgrOQljYzAcv0(a66EQGu3lF5u04PjOrwycDfHd1IPPGV(9dlbcMnxDbs2ZUyh9RFGMqjW5T3sVAPYMNwfndkIpYMAknu556UETn521xAMAdorQn(ndqZMbjLYj(xmWRQz0yQbfJ3vdM0sVUPz9f3jxAbMArDw5Suth0MPN4i)nfWVWFCbR4fyV4dqchM2APyrL7sWLkwTWInmRRxTqXNGwnvmRSCjRUCHWxGT(aFxRJCImq0Q2Capy2J7QlqP4DnRUcJSwb0KiC1g(BYeXdxKr5Kvtwv9CgvR5PG6yY4yO(BW79qa2wWKSwhxBQM6hn4wf3U6YpB2PlEF)KE1ZScy0n(ZN(3p


Then the same auras, after the sub element tweaks:
!WA:2!LA12YnTryy8OHHIgkKeGuowCdfMKs5qCOHwAPTwo2eN4y7kRKaDshZkTRTwIS0MvRYP7WddtUopc(AUYpc5jyNm9jipc5UEx)xj7CGPq7KEJ8EE)))((()wNAHHcZmfz1HAneEi8w3CxBKZYyEalxGxaFgnnTQAp6bF3E2bCmHB0Bs2agE0n3eXXPTcc8euwhrGZQeEinW)kPQ)xMHoipIMDY2C5BuPrJqIivhoPjSeRnyeZM8GiwNKvuJUj50DtANCZPsLsBxuKWnGxHjG9eQ3f57aDRgq9f25Yx2kVjh5ep3KMHcexOBq9PcD7gWpHU6TfCAZMquDT7Y718DDWe7OgnuraF68LQwy(sgcv4GI4OmTdzepVI4q99cJSjRs8f1GvtxVB9CzRzvVMvwtlJi4sSzEOniCtFulsOUz8sTNMG8eUhS1QCcSvZAvZxQ02r(9cb9E5CECtcBWAReH4K0fI88sVOlvq2jz2ea7S817HCBhs8AeN6YlmcfVDzYAPFUcbn8cq4cgHacoSzRiGm0nGSWzy591nD8qHHQw2cGp8fQMDBWHyUMGJeiT9tW0cQHuyIDTCM5ZxEF4EiCFK3cjCkQh3u0hIfTTr(0wifWpPCWcYRy4h4t2HGcvNkXVPW9m7Ib4uTI6kWLhsCc8XHTvRrDnYNy0cr9H9kFI87pJ8hKpv(JqZl(HJ0bWsZyntOUnCgnOn17OolApnb1Vrapjy0ZerXD9gnALXgRyKZSMd5mZwJ0HclVwqe3HCJtjhnL8ZsjpV8c76acxCWA(1wJYiVAN(DvSIRCaiX0LdPlVOU8sN)UYl)o5WYVaITRkVMU86YBiVP8lL3sxMw(vYrK3w(16SbvAXvjwjS8CbyY7pL8XO9xMqyzvmIWufMUYXKxauHHiHcIGl7zFZWYRSlKOQApvvHpggqw02LqB6kEJ8EfKFRI5Kpi(7dJ)(O4VN1qLEYpxVD)eWvozcRCCK9N(3XAJndcALYe5XCrAYtRQJbLwgqPz603nqtoUMmJCc5pRlZ2DC)8R2kZJEnE8YCQVQ(N4YUu)iXISUykAiY2JGDnxJIfUVr(l6YFvxo7R26HDczi(YlQg(YYChJBINXmqetRPydES(kK1i78wv2)q86apJy3Q)RuMSW7pvS8O(rYVhJoXayBBup3lnW9soT7oXbEsf8ltjF2tJ51ESnOBjhY2)HCMB9ji69IpjdOggRY)mzNAk2arHKS4xhfkiycEo06UkYzw3pCckOjolx5VbX3rL2hn8ErQTJrIqkMyAw85tBbIk5CP6gVMPJvIpJlaknItA33(pHdp4TcTKt0kzvSrlQCrAGCilLdfkO(nnq8LMV49pS39RftTXBZnrZ17gPymX3SC(fYBgl4kRjTE75abPPY6lJsaA6cAWf9SIMS1knMA(x3nGtbtUyvsNPRyw83Ru2kBPEAoOSBR7a4TWJ6d(RZBvQy58GOJDpmnu5MRc76jUj1z1f0wK64na)DQtDHlNe6cY6n2n0fbI7xMGAGQf0RS0F8tiPxk2aoaDe0sTI5zyOUxMEp45TaL1LtCrGrwO7jX5rjJzx)FmesAyQIe8)hLTrJaFbBGcC6MP)Tiew92rAlReRRyfTS4Nq92wTD1B8xTh89Ie4td8VusQUhj4h5om2T)44jJtCOQNK0KvJLCDGNyj4ye8TNJRe90gByukFbRdSWIvmt0En4jSf5iw7f71OVxg46eBN55Ts51OovMFIYG4VBsOE4FhPVoY2OILvL52kDp7QtkNPiequc(eT5996ooIoZBpND8uKx9jq397vvQsjL5a7s5kLDUQwvmkLn3SGtrrRIlK3UfLZd4GvGS5229nsmbKQ8uYMSjpSwnlgdpTU0Ie0YzbtJWLQgSgHh3CocMIwkzu13juq4jvuDC(5Xk)MYjescZcmYy52S0cVCYXZ9Kav1pBqqy4Wjcs9drmG6siMrtn0Q)5l(7)

And last, what makes it complicated. A user adding new sub elements to auras that contain sub elements, here:
!WA:2!LE1wZTTXvyJH1tQ5KeljFjY5QSsTh7j12YuXs2T1TLqIYI20KYGqs2UQLAjWscybcaVaq3AY0yoE8Oh7OoD67SVMxc)jOP3tBZPSPPVRjt)bO36B9S7cqs5XY1JZlG7LZU7zp73335qLfgkiZ00vhQXqMdzU170PkXyftMN)uEoESBMkvQ5sn2fVYUv9yMuMA8K(dO6yV5MeM5i6EEoH2(Tc9mwLYcS9ChwPY)vlWG4qtvvUml2gLQvlGgQ0IrRJMOVHpvRoZlYVL0IY2BspCBzB5jROOKQdjk0YJvYpextq62exdS7CE2UHvNkxr9CAmIHyUj0ccjSW0Q2U2HPRwd)jWkDZqMD96Ox9MNLf38jTmPvJQvJ7bSzZvyUzMVGAi3DirmsMMb(uhN8MbP3niQkDvQByz0A71BxzQSL1RuwpRMUAeEiv9DiBqzAUKg0G0ActRolL4eA1DPZXO4s1kpxUcf2oYn2fshFNZzwN6py5hgry0rMjYXzKfTSdP7iNvgWocB94i32buNAIRoC0rTn3UiDTrUbpcQ64rmNrnaJGNuRre(yKwfVfgNeUqAndhsqaVv1q89WnK3SDng6ZLdzKqsQ9KX0z4dXJjvlpLwUCf3dphkZL4SG8nLe)2K3f9LuBtCTBq4b(jGbNbgw11ZLUdLeW3vQB9qRxPJjgo5wuHhCzbudpxZGMCB4hdmPAdITlUwys4QVcCn47bFFS5XE6rAHXsnbMjiDvCpQzxpDl(EzhJjSDR5XKot6mr2MTDox0dp)5ZhzClTHmU5wJ2YgnVSxeZG(2hcoNc8TvGxhoAhde4A6TMB51S9PlVtsx(RIfmaEXsddLgowA44V(zHt8e4KWBG(2PG3mn8wWBdVd8UW7Lggbonmk8(W3jT)GCS4QuD5R8T9mPF6HGpKS3kuQFw(lsOg3nTGZdhfrHbKqEicpSRF)VQtDhV1MHrFye11yJ5(wJDXmxbgMHxE(e7efqVb(RGzy1Kpepi2PAuyONBjK3HyXw8HliI(NOnV99sqo1twkNoNsL31syXD73IYcgRWr0TSnwXLgeKsmtbBS9GI9xvYNpjmCh034AfCwSR5wVH)PdPRhwr8r(IuXVsODdAfFg1WMdIsTDxtg9mOgIOxGfXmX94C(DLJMa2vZoVEj)Jjgevd84V0glABgAPMf7UN88SxNAkg8rDK45K7(oI5FquqODTnWWoZled564GQflvmNmeZTPLWqJUrPTJViUH(dmdZEZrUteXKZygrxxUPRI3PQo0L9hQV7rY5g7gjb4octwddElYi(nxmUrNUhIqoQ086fYxmxCGPpXo4O(V7bfDL9u8p1ZWazdnUDM9olUw7P8p)bTHMBGsA2gvcTy0alKuSX(UHjxjecalwdgEBooqGkILiQWvIoCSkwFr04uiDJBjMtrchC2oX98K7EvlQDDRWhbFWmW3LlAbxu89sIVJj(EevoZgET0nt4UwWesbP9lQ8d()lZOUPNxdfnIJVfjfCyEkmuKndkYQ17gaxofKbgh(HPHSTVSBUvBKzShyE5ImBxEQpQL)Xt8eocBA7acErnT0wtanHFuA4hNg(4L36sTc8jSvei2tatTpzjXmACykYyu8hCF95IkWs71J31L3iYs)IkvbZ8Phsil23ZdQu9sh9AwLeN1ofYGHzT2r41sKY9uGRV5)bYNiQb3eULfuqBoK06afpbusbMt6c3Xc0tbLtjeNIjazW9g0uG5hewWsa6wf)SAcErAe8lH1omSUCx2yzyt4SWpxjwLcH30yvQJdrPG7n6zCH7leBGFcSe8tHFgu5rWYkabokufmW93eOYnRguhSXhOvahOHArmph41oFXI50QOL)gZQdwldbNcEWZIa6g3aMK73BV1a7HAxRubXeUMvAGpLzYo90Xsq4e7iMnjFV02(st4pqx9Vkocz(L3vytyIA9HrDs(ayLjg0gy2EXHWYgUaXjI2gtIiCVig1QLWWEV(YZEJUQMI9vAR)5YZRfOgUNlnfbfsDRRsylnF(l0R3fkZFVLllHFV7(9LrVYyTfJi99Rl70WMXWuADeDyjOE4JEVNdXFxb4sLhff8HFJ)a4vlRjxLNAsnVnzDlov9JTE6jSrfIJW4f6Xbv9LJVF86Dv2wqncSnPAY34xln8luAlSzwHU01zXbNMj1blz09YYk3rDPvWVvt01sQSeVt2MMuxTI5wiNMqw5tsbp5XVkk7OXRTldxM5XwOsZIo6rt04H1ME(h02JzJXsruQ1SL0YF)sf1ZwiwzbRRyRZGa9qhmHn4IIk(FGPDaVqv9xeD(o7lhgc6r9i)ro4Dij1Zagya2Rb3I59nrSkmYU7pxnYWEjQOIlt5)wpZZpMyX(MOAPYtgI07RF)TgkdX32AxeviOmovmCS9xUTSDqyLOG70RdBtfy4MiPvmYoYXRz7qZBwnZ4xz8XVAl5y(Kql)tl(hfbxIWmiUuFV14zi5fHuXWIgeEXgz6TZHBOWKD2qPVZB9VEGMD5gXgSzFgeUE)w3DPYKzDKD6M1T3Q2mzVwxMmFRHGFNf87xg(d3b(J4ve(tWFU61M4AtETRcFU)zIVhtuzSkRrj(Osik9sqrglugGwbvSX7c83uG)Uc8fF9aaOa)dfOJc8pvGVmf8Vwg(kf4F)8oQjhBSRnbEuN9apkc6XSkwif8f6SEoIin7wjuN9vxtkSSco3UDFWoEXQV)btd6vFl0uW9B1RE0h)QS4IpvlKBg9UvwiO4J3mPIqm5sCTfyvaIYlCCEyX1Snkn)4fr5N2pDvYjm(QQL01lD7TgjU8HxwAgNhHYhyQBESezeFdZuc)kWRL03e37KSKWVwMmKx5iS6bK0(zJpG)c85WF9f4r)JE8RwvwN)YphaWEXk48WppnI)XNQq2BpNEj1czN6wyI5865xix14euWrGpB7QjPC0WhZItdFM)e9snM10e))OlTiLSswmbtWsZXP6IM3MAAtwsok)748h7xwvR9dH(qEUPprcDKGpe7C(P2SWc3BIlp1KE8mk(dIyxdgfjP9YWIGmje6CkdT6xE3)3d

You probably should add a few conditions to these auras for a in depth test.

We also should remove all the sub regions from the default tables, and probably take another look at the ValidateSubRegions code I had previously written.

Reply:
good catch on indexes[i], i'm stupid